### PR TITLE
Add cached `run_system` API

### PR DESCRIPTION
**Pull request by benfrankel**
# Objective

Working with `World` is painful due to lifetime issues and a lack of ergonomics, so you may want to delegate to the system API. Your current options are:

- `world.run_system_once`, which initializes the system each time it's called (performance cost) and doesn't support `Local`. The docs recommend users not use this method outside of diagnostic use cases like unit tests.
- `world.run_system`, which requires you to register the system and store the `SystemId` somewhere (made easier by implementing `FromWorld` for a newtyped `Local`, unless you're in e.g. a custom `Command` impl).

These options work, but you're choosing between a performance cost and an ergonomic challenge.

## Solution

Provide a cached `run_system` API that accepts an `S: IntoSystem` and checks for a `CachedSystemId<S::System>(SystemId)` resource. If it doesn't exist, it will register the system and save its `SystemId` in that resource.

In other words, it hides the "save the `SystemId` in a `Local` or `Resource`" pattern as an implementation detail.

Prior work: https://github.com/bevyengine/bevy/pull/10469.

## Testing

This approach worked in a proof-of-concept: https://github.com/benfrankel/bevy_jam_template/blob/b34ee29531e3ceae287a9cc44ec6c520e83a4cdd/src/util/patch/run_system_cached.rs#L35.

A new unit test was added and it passes in CI.

### CODE REVIEW

**Comment by benfrankel** [comment]
Note: Now `run_system_once_with` is the odd one out in the sense that it accepts `input, system` in that order rather than `system, input`. This can be changed in a separate PR.

**Comment by janhohenheim** [comment]
Great addition! Code looks good to me, but can we make the docs a bit more approachable? I understand them, but I doubt the average user will. Why is this useful? When should I prefer the uncached vs the cached variant? Maybe mention Local, and EventReader in particular. Further, can we link this from other places? I recall some part of the docs saying "running a system like this will not cache anything", maybe that can link to these?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Seconding the doc changes, and also requesting that this be a full replacement for the existing run_system API. The ergonomics of this are *so* much nicer, and we don't need yet another variant.

**Comment by benfrankel** [comment]
> Seconding the doc changes, and also requesting that this be a full replacement for the existing run_system API. The ergonomics of this are _so_ much nicer, and we don't need yet another variant.

I think making this the _default_ API (aka shortest name) would be reasonable, but I'm not sure about fully removing the existing API. Sometimes you _do_ want to store and pass around a type-erased identifier for the system, like in an `OnHit(SystemId)` component. So you need to be able to run the system from the `SystemId` somehow.

Removing `register_system` in favor of `register_system_cached` makes sense, but then `Commands::register_system` would have different behavior than `World::register_system`. The former doesn't know whether it should spawn an entity until it checks `World` later, so it has to either always spawn an entity (not cached) or wait, and not be able to return a `SystemId` as a consequence (which mostly defeats the point).

**Comment by alice-i-cecile** [comment]
Let's swap it to the default API for now then. I wouldn't mind removing `Commands::register_system` completely (you can just write a custom command...), but I'll see what others think.

**Comment by janhohenheim** [comment]
Completely agree with @benfrankel here based on my experience.

**Comment by vil-mo** [comment]
I find it strange that `S` and `<S as IntoSystem<I, O, M>>>::System` are cached twice, even though they cache the same system. You could replace the resource type with
```rust
pub struct CachedSystemId<S: System> {
    /// Cached `SystemId`.
    pub id: SystemId<S::In, S::Out>,
}
```
This will also allow the type level to more clearly define what the resource does and get rid of unnecessary generic parameters.

**Comment by benfrankel** [comment]
Wrt making `*_system_cached` the default API, it's not clear what should be done for the `register_*` variants.

1. `register_system` and `register_system_cached` have the same API accepting an `IntoSystem` and returning a `SystemId`, but the latter also saves the `SystemId` in a resource. Should `register_system` be fully removed and replaced with the cached version? If not, how should it be renamed? `register_system_uncached` doesn't seem great.
2. `register_boxed_system` accepts a type-erased `BoxedSystem`, so there can't be a cached version with the same API. Can `register_boxed_system` just be removed? If not, should it be renamed to make it clear that it's not cached like `register_system` would be post-rename, or is it sufficient to mention it in the docs?
3. `Commands::register_system_cached` doesn't exist atm because it can't check the `CachedSystemId` resource until the command runs with `World` access. As such, it can't `.spawn_empty()` ahead of time as there may be a pre-existing system entity, so it can't return a `SystemId`. Should it exist anyways but with no return value? And should `Commands::register_system` be renamed to `register_system_uncached`, or removed?

This PR is blocked on these questions. If there's no good way forward to make `register_system_cached` the default API, it may be best to keep the current names.

**Comment by janhohenheim** [comment]
My two cents:

1. After some reflection, I think the non-cached version should be removed. Using that over a command is too niche.
2. No idea whether it should be removed or not. I *think* mentioning this in the docs is good enough? Otherwise, may I suggest the name `volatile` for `uncached`? It directly suggests that this API brings some intrigue with it.
3. Is `register_system_cached` even usable without a `SystemId`? How do you retrieve the system? By registering it in `World` again? That sounds rather pointless. I would probably keep only the uncached variant and rename it.

These are not strong opinions, feel free to disregard.

**Comment by benfrankel** [comment]
Okay, here's what the full proposal would look like along those lines:

| Type | Method | New name |
|-|-|-|
| `World` | `register_system` | `force_register_system` |
| `World` | `register_boxed_system` | `force_register_boxed_system` |
| `World` | `run_system` | `run_system_by_id` |
| `World` | `run_system_with_input` | `run_system_by_id_with` |
| `World` | `remove_system` | `remove_system_by_id` |
| `World` | `register_system_cached` | `register_system` |
| `World` | `run_system_cached` | `run_system` |
| `World` | `run_system_cached_with` | `run_system_with` |
| `World` | `remove_system_cached` | `remove_system` |
| `Commands` | `register_system` | `force_register_system` |
| `Commands` | `run_system` | `run_system_by_id` |
| `Commands` | `run_system_with_input` | `run_system_by_id_with` |
| `Commands` | `run_system_cached` | `run_system` |
| `Commands` | `run_system_cached_with` | `run_system_with` |

Notes:
- `remove_system_by_id` can't remove the `CachedSystemId` resource, but that should be fine because the "invalid cached ID" case is already handled by other methods.
- The current `Commands` doesn't include a `remove_system`, presumably because it can't return the removed system immediately, which is similar to why a cached `register_system` won't be included.

> Is `Commands::register_system_cached` even usable without a `SystemId`? How do you retrieve the system? By registering it in `World` again?

You can't exactly use `Res<CachedSystemId<type of foo>>` as a system param, but if the `CachedSystemId` resources were consolidated into a single `struct CachedSystems(TypeMap)` resource, it would enable accessing that resource and calling some `CachedSystems::get<S>(system: S) -> Option<SystemId>` within a non-exclusive system. Exclusive systems can just do `world.register_system(system)` instead. So this may be something to consider.

**Comment by janhohenheim** [comment]
@benfrankel I quite like these renames :)

**Comment by benfrankel** [comment]
@alice-i-cecile requesting feedback before I do a rename, because it'll touch a lot of files. It may also be nice if that can be done in a separate PR so that this one doesn't become a confusing new-functionality-plus-refactor mega PR.

**Comment by alice-i-cecile** [comment]
I really like the rename, with one nit: `force` -> `manual`, and move the modifiers to the end of the method name. I think that this is clearer and more consistent with Bevy's API. It also makes the methods appear beside each other in alphabetical order, which is good for docs and autocomplete.

**Comment by SkiFire13** [comment]
I believe this will suffer from the same issue I mentioned here https://github.com/bevyengine/bevy/pull/10469#issuecomment-1804113209

```rs
#[derive(Resource)]
struct R(i32);

let mut world = World::new();
world.insert_resource(R(0));

let make_sys = |n| move |mut r: ResMut<R>| r.0 = n;
let _ = world.run_system_cached(make_sys(1));
let _ = world.run_system_cached(make_sys(2));

assert_eq!(world.resource::<R>().0, 2);
```

With this PR the assert fails due to the resource holding `1` instead of `2`. The issue is that `run_system_cached` considers the two systems the same because they have the same type, while they are actually different because they have a different internal state.

The worst part is that there's no warning or error with doing this, so this is only observable through its wrong behaviour, which is horrible for debuggability.

**Comment by benfrankel** [comment]
There's also the issue that function pointers with the same signature have the same type, like `x: fn()` is the same type as `y: fn()`. I think it would be a shame to lose a useful API because of some edge cases like this, but it is a real issue. One option would be to document it and call that good enough, but I'll give this some more thought.

**Comment by janhohenheim** [comment]
I think documenting it and hoping for some fix in the future is good enough for now. People running into these edge cases can still use today's manual API.

**Comment by SkiFire13** [comment]
> People running into these edge cases can still use today's manual API.

IMO the problem is that:

- people most often don't carefully read the documentation, so they can easily miss this;
- there's nothing at runtime helping debugging these issues.

When combined, you get people that have no idea this issue could happen and have no help detecting where it comes from. Thus a warn when `std::mem::size_of::<S>() > 0` seems sensible to me, since that would at least catch all cases where there would be an issue. I would however prefer something even stronger, like a panic-by-default that can be disabled with a feature flag, so that it is clear that this is something not supported that can lead to very weird behaviour.

Ideally we would be able to compare any system to check if they're equal and cache using that, but it's not possible due to `Eq` (and `Hash`) not being implemented for neither function items nor closures. Note that even if they get implemented in the future it will require a breaking change for these methods.

**Comment by janhohenheim** [comment]
> std::mem::size_of::<S>() > 0 seems sensible to me, since that would at least catch all cases where there would be an issue. I would however prefer something even stronger, like a panic-by-default that can be disabled with a feature flag.

That's fair. I'm very against adding more feature flags because of combinatorial explosions, but I agree with the general sentiment. Maybe add an `_unchecked` variant instead? Or just disallow that case (`size_of::<S>() > 0`) completely and only add that piece of API if people really need it in practice.

**Comment by benfrankel** [comment]
In order for this PR to move forward, I think the following will need to happen:

1. Deny closures and function pointers by requiring `std::mem::size_of > 0`.
2. Document the limitation.
3. Consider whether the cached API should _not_ become the default API for running one-shot systems due to the limitation.

**Comment by janhohenheim** [comment]
Those limitation sounds alright to me. Pretty sure it should still be the default though, as those are niche uses.

**Comment by NthTensor** [comment]
I like this direction and I support this work. Getting this working for just concrete functions will be enough to reduce some of the dubious stuff being done with global observers.

I am tentatively in favor of this becoming the new default, but would like to see it merged before we commit to that.

**Comment by janhohenheim** [comment]
If this is merged before becoming the new default, I'd like to at least heavily point to it in the documentation of `run_system_once` and `run_system_once_with`.

**Comment by alice-i-cecile** [comment]
> In order for this PR to move forward, I think the following will need to happen:
> 
> 1. Deny closures and function pointers by requiring `std::mem::size_of == 0`.
> 2. Document the limitation.
> 3. Consider whether the cached API should _not_ become the default API for running one-shot systems due to the limitation.

I'm on board with all three of those. IMO we should make this PR non-breaking (no renames) for now, given the caveats that exist.

**Comment by benfrankel** [comment]
Okay sounds good. I'll work on updating the PR in that direction.

**Comment by benfrankel** [comment]
Denied `size_of == 0`, improved docs a bit, and added a unit test. Ready for review again.

**Review by SkiFire13 — APPROVED** [review]
The PR looks in a pretty good shape to me.

The only maybe issue I can see is that the ZST check won't work for systems created by `pipe` and `map`, even when the systems and closures involved are ZSTs. This is due to `pipe` and `map` eagerly calling `into_system` on their receiver, which will result in non-ZST types even when called on ZST types. The solution would be changing them to be lazy and call `into_system` on their receiver only when `into_system` is directly called on them.

**Comment by alice-i-cecile** [comment]
> The only maybe issue I can see is that the ZST check won't work for systems created by pipe and map, even when the systems and closures involved are ZSTs. This is due to pipe and map eagerly calling into_system on their receiver, which will result in non-ZST types even when called on ZST types. The solution would be changing them to be lazy and call into_system on their receiver only when into_system is directly called on them.

I'm happy enough to call this "fix in follow-up", but this is a good find.

**Comment by janhohenheim** [comment]
> I'm happy enough to call this "fix in follow-up", but this is a good find.
Agreed! Could one of you (either @SkiFire13 or @benfrankel) write that down in an issue?

**Comment by benfrankel** [comment]
Made an issue: https://github.com/bevyengine/bevy/issues/15373

**Review by alice-i-cecile — APPROVED** [review]
I'm happy with the API now. Thanks for exploring this so thoroughly.

**Comment by alice-i-cecile** [comment]
Really lovely work; thanks for seeing this through.

**Review by NthTensor — APPROVED** [review]
Sick! I do think as follow up we should consider flipping this around to be the "default" oneshot api.

**Comment by kaphula** [comment]
Hello, 

I am testing the latest main branch of bevy now with this feature merged, but I am running into the compile time assert check for closures and function pointers even though my system should be normal?

My system definition looks like this:

```rust
pub fn start_road_building_system(
   gr: Res<GlobalResources>,
   gs: Res<GameState>,
   mut players: Query<(&mut PlayerStateECS, &HouseECS)>,
   junctions: Query<(&JunctionECS)>,
   road_paths: Query<(&NewRoadPieceECS)>,
   roads: Query<(&NewRoadECS)>,
) { ... }
```

And I am first using `world` instance to register the system as cached, then in another system, using `Commands`, I am calling `cmd.run_system_cached(start_road_building_system)`.

Is there something I am missing, should not this work?

**Comment by SkiFire13** [comment]
I think this is a bug in how the `RunSystemCachedWith` command is implemented (it stores the result of `IntoSystem::into_system(system)` and passes that one to `world.run_system_cached`, but that is not a ZST). I'm working on a fix, as a workaround you should be able to use `cmd.queue(|world| world.run_system_cached(start_road_building_system))`

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1690 if you'd like to help out.
